Source/CarthageKit/Git.swift

```diff
@@ -269,16 +269,16 @@ private func parseConfigEntries(_ contents: String, keyPrefix: String = "", keyS
 	}
 }
 
-/// Git’s representation of file system objects at a path relative to the repository root.
+/// Lists the contents of the directory at a given path relative to the repository root.
 ///
-/// - parameter path: Path separators at the end of `path` have significance of outputting directory contents.
-///                   Thankfully, multiple contiguous path separators seem to have no adverse effects.
 /// - note: Previously, `path` was recursed through — now, just iterated.
 internal func list(treeish: String, atPath path: String, inRepository repositoryURL: URL) -> SignalProducer<String, CarthageError> {
+	// git ls-tree treats "dir/" and "dir" differently. This makes sure the path is a directory
+	let directoryPath = URL(fileURLWithPath: path, isDirectory: true).relativeString
```

Comment:
This is a tricky corner where we’re passing paths to check git’s storage in anticipation of writing symlinks at those paths (or filtering those paths out).

Every choice here works well enough for [the single call-site](https://github.com/Carthage/Carthage/blob/0f561ee/Source/CarthageKit/Project.swift#L951) where we invoke the function, but I’d prefer to leave it dual purpose: listing single files as well as contents of directories.

> This comment could definitely be improved — the dir and dir/ examples are good visual indicators.

In that light, we should avoid NSURL here when appending that trailing path separator we’re looking for — just `path + "/"`¹ at [the call-site](https://github.com/Carthage/Carthage/blob/0f561ee/Source/CarthageKit/Project.swift#L951).

<details>

<summary>
And, there reason to actively avoid NSURL here: <a href='https://en.wikipedia.org/wiki/Null_character'>Null characters</a> with `URL(fileURLWithPath:)`
</summary>

```
  7> URL(fileURLWithPath: "etc\u{0000}etc", isDirectory: true).relativeString
$R5: String = ""
  8> URL(fileURLWithPath: "etc\u{0000}etc", isDirectory: true).scheme
$R6: String? = nil
```
</details>

<br>

Unless there’s something I’ve overlooked, what do you think?

⑴ `path` being `carthageProjectCheckoutsPath` here.

Reply:
I think the git ls-tree behavior of treating `dir` and `dir/` differently is somewhat surprising, so maybe avoiding it is better than exposing it to the user. The main purpose of ls-tree is "list the contents of a tree object" as the git document states. Passing a file path or a directory path without an ending `/` such as `Carthage/Checkouts`  just gives you the path `Carthage/Checkouts` back, which does not make much sense, unless you want to check if a file or a directory exists. But in that case, I think it's better to use another function for checking the existence of a file since this is naturally a different usage of ls-tree.

I don't think it's a valid path string with any null character in it. This kind of error should be solved before passing the arguments to this function.

Reply:
I think roundtripping through `URL` is unnecessarily complicated here. The Path<->URL conversions in particular are not very well documented, nor do I think they're very well understood. (I definitely don't have a mental model to understand the safety of this operation.)

Given this comment:

> Thankfully, multiple contiguous path separators seem to have no adverse effects.

I think the safest thing to do is to append a `/` to the incoming `String`. Since that's the behavior we want anyway, let's just do that.